Mobile table v2

youngfreezyVA opened the pull request:
here it is in action:
https://drive.google.com/file/d/1GwDAfr5TvHeid0t_DuNp4hKQMpL4iruH/view?usp=sharing

https://github.com/department-of-veterans-affairs/va.gov-team/issues/15375

Commit:
save progress so far

Commit:
save work so far

Commit:
save work so far

Commit:
Merge branch 'master' into mobile-table-v2

Commit:
get it working

kelsonic commented:
Could you ping us in a comment here when it's all ready for review? 🙂

Commit:
rename filter for clarity

andaleliz reviewed (commented):
:clapping: looks great! The only thing I see in the video is on the desktop version - the cell data should be top aligned instead of center/middle aligned. 

I didn't review the CSS, focused on the look and feel from the video.

youngfreezyVA commented:
@kelsonic @jenstrickland @ncksllvn this is ready for review!
 @andaleliz thanks for taking a look and the feedback! I didn't touch the desktop headers :) didn't change the html markup for desktop, and all my css changes are wrapped in a media query.  if that's an issue we can create a separate ticket for it and i'm happy to take that one on

kelsonic commented on src/site/paragraphs/table.drupal.liquid:
```diff
@@ -27,31 +27,37 @@
+                        {% endfor %}
+                    </tr>
+                </thead>
+            {% elsif forloop.last == false %}
```
Won't this make it so that if there is more than length 2 for `entity.fieldTable.value`, the middle values won't be rendered?

kelsonic replied:
Wondering if that should be an `else` instead?

kelsonic replied:
Oh nevermind seeing now that this is how it was previously, so better not to mess with that lol

youngfreezyVA replied:
hehe agreed :)

kelsonic reviewed (approved):
LGTM! 🎉 Nice work and perseverance with this one!

ncksllvn commented on src/site/paragraphs/table.drupal.liquid:
```diff
@@ -27,31 +27,37 @@
-                        <td>{{ column }}</td>
+                        <dfn>
+                            <td class="vads-u-padding-bottom--0 column-label vads-u-font-weight--bold medium-screen:vads-u-display--none" role="cell">
+                                {{colLabels | getArrayItemByIndex: forloop.index0}}
```
I haven't seen `forloop.index0` before. Are we just trying to get the first element out of `colLabels`?

youngfreezyVA replied:
`index0` is zero based iteration whereas `index` starts off at 1... smh drupal :P
we are trying to get the associated col label (From what used to be the table header, which is hidden in mobile) for each column value

ncksllvn replied:
Wow yeah, I see it in the Shopify docs now, https://shopify.dev/docs/themes/liquid/reference/objects/for-loops. Thanks for the insight!

ncksllvn commented on src/site/filters/liquid.js:
```diff
@@ -368,6 +368,10 @@ module.exports = function registerFilters() {
 
+  liquid.filters.getArrayItemByIndex = (arr, idx) => {
+    return arr && arr[idx] ? arr[idx] : '';
+  };
```
There is already a [`get`](https://github.com/leizongmin/tinyliquid/blob/cebcb26c3839e725cf0469dccc0073799902a020/README_en.md#arrays-and-objects) filter that might do what we need? It works to get dynamic keys out of object, thinking it might work for array indexes as well?

youngfreezyVA replied:
let me check it out...

youngfreezyVA replied:
you are correct ... `getValueFromObjPath` works for arrays as well... so I added in a comment above `getValueFromObjPath` to indicate it works for arrays also

ncksllvn replied:
Oh cool, thank you 😄

ncksllvn commented on src/platform/site-wide/sass/modules/_m-responsive-table.scss:
```diff
@@ -14,4 +14,50 @@
+  }
+  .rwd tr td:nth-child(3) dfn {
+    display: block;
+  }
```
What are your thoughts on `va-table--responsive` instead of `rwd`? 

I'm also wondering on the "3" in `td:nth-child(3)` - what that numbers means versus a 2 or 4

youngfreezyVA replied:
i tried to add in va-table--responsive instead of rwd and it broke everything 😬
the `rwd` class is from jens codepen
re: the specific selector you pointed out.. it wasn't doing anything, so i removed it as well as a couple of other selectors that were specific to the structure in the codepen and didn't match the data returned from the cms.

ncksllvn replied:
Oh I just meant if we could use something like `va-table--responsive ` as the class name instead of `rwd`. Off the bat, I don't know what `rwd` stands for and just thinking we could use a more descriptive name

youngfreezyVA replied:
ah gotcha... that makes sense - done.

jenstrickland replied:
Yes! Y'all should definitely change `.rwd` to use the naming convention used on VA.gov. I used `.rwd` in mine because I don't know what y'all would use (rwd stands for responsive web design, but it's totally my shorthand sketch).

ncksllvn commented:
@youngfreezyVA Thank you for the fantastic video btw!

Commit:
update w/ Nicks feedback

Commit:
Merge branch 'master' into mobile-table-v2

Commit:
Merge branch 'master' into mobile-table-v2

Commit:
add back css changes -- regression was from cms data structure change

Commit:
update class name with Nick's feedback

Commit:
Merge branch 'master' into mobile-table-v2

Commit:
add back elsif

Commit:
Merge branch 'master' into mobile-table-v2

ncksllvn commented on src/platform/site-wide/sass/modules/_m-responsive-table.scss:
```diff
@@ -14,4 +14,40 @@
+  }
+  .column-value:nth-last-child(1) {
+    padding-bottom: 20px;
+  }
```
This is looking great. I'm wondering about the `td` rules and the `.column-value` rules. Should they also be prefaced by `.va-table--responsive`? 

Your call, but we could just wrap the whole thing ->

```scss
.va-table--responsive {
  tr {
    border-top: 2px solid #999;
  }
  thead {
    display: none;
  }
  td,
  th {
    border: none;
    display: block;
  }
  td {
    display: flex;
    grid-gap: 2px;
    padding: 1rem;
    padding-left: 0px;
  }
  dfn {
    display: block;
    font-style: normal;
    font-weight: 800;
    margin: 0;
    padding: 0;
  }
  td p {
    margin-bottom: 0;
  }
  .column-value strong:nth-child(1) {
    font-weight: normal;
  }
  .column-label:nth-child(1) {
    padding-top: 20px;
  }
  .column-value:nth-last-child(1) {
    padding-bottom: 20px;
  }
}
```

youngfreezyVA replied:
💯 yes i thought to do the same thing but forgot. good call

Commit:
add comment for clarity around elsif

Commit:
wrap with parent class

Commit:
Merge branch 'master' into mobile-table-v2

ncksllvn reviewed (approved):
LGTM! Thanks for making the updates and so quickly! Just please test one last time before merge to make sure everything is 👌

Commit:
Merge branch 'master' into mobile-table-v2

youngfreezyVA commented:
> LGTM! Thanks for making the updates and so quickly! Just please test one last time before merge to make sure everything is 👌

yup its looking like everything is working as expected @ncksllvn ... do i need to wait on anyone else to merge this in?

ncksllvn commented:
@youngfreezyVA Yeah I suppose we do need to wait...If you look at the `Code owner review required` we can see that the frontend-review-group (this is the platform team, aka VSP) still needs to sign off

billfienberg commented on src/site/paragraphs/table.drupal.liquid:
```diff
@@ -27,31 +27,38 @@
+        {% assign colLabels = entity.fieldTable.value.0  %}
+        {% for value in entity.fieldTable.value %}
+            {% if forloop.first == true %}
+                <thead>
```
I think rendering the `thead` as a child of `tbody` is technically invalid HTML. 

>4.9.1 The `table` element - Content Model: In this order: optionally a `caption` element, followed by zero or more `colgroup` elements, followed optionally by a `thead` element, followed by either zero or more `tbody` elements or one or more `tr` elements, followed optionally by a `tfoot` element, optionally intermixed with one or more script-supporting elements. - https://html.spec.whatwg.org/multipage/tables.html#the-table-element

>4.9.5 The `tbody` element - Content Model: Zero or more `tr` and script-supporting elements. - https://html.spec.whatwg.org/multipage/tables.html#the-tbody-element

youngfreezyVA replied:
addressed this feedback as well :)

billfienberg commented on src/site/paragraphs/table.drupal.liquid:
```diff
@@ -27,31 +27,38 @@
+                <tr class="vads-u-padding-top--5" role="row">
                     {% for column in value %}
-                        <td>{{ column }}</td>
+                        <dfn>
```
I think rendering the `dfn` as an immediate child of `tr` is technically invalid HTML. 

>**4.9.8 The `tr` element - Content Model**: Zero or more `td`, `th`, and script-supporting elements. - https://html.spec.whatwg.org/multipage/tables.html#the-tr-element

I think it would be valid if you moved the `dfn` to be a child of the `td` instead of being a child of the `tr`.

youngfreezyVA replied:
done

Commit:
Address pr feedback

Commit:
address PR feedback

Commit:
address PR feedback

youngfreezyVA commented:
@billfienberg @kelsonic @ncksllvn @andaleliz 
addressed all feedback and here's the responsive table in action, post pr feedback:
https://drive.google.com/file/d/1hoaMSfFHFiho6KosKns78K9tL6EMnKHN/view?usp=sharing
